<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="GOTTCHA : Accurate read-based metagenome characterization using a hierarchical suite of unique signatures">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>GOTTCHA</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/LANL-Bioinformatics/GOTTCHA">View on GitHub</a>

          <h1 id="project_title">GOTTCHA</h1>
          <h3 id="project_tagline">Accurate read-based metagenome characterization using a hierarchical suite of unique signatures</h3>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/LANL-Bioinformatics/GOTTCHA/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/LANL-Bioinformatics/GOTTCHA/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
<section id="main_content" class="inner">
        
<h3>
<a id="genomic-origin-through-taxonomic-challenge-gottcha" class="anchor" href="#genomic-origin-through-taxonomic-challenge-gottcha" aria-hidden="true"><span class="octicon octicon-link"></span></a>What is GOTTCHA
</h3>
<p>GOTTCHA (<u>G</u>enomic <u>O</u>rigin <u>T</u>hrough <u>T</u>axonomic <u>CHA</u>llenge) is an application of a novel, gene-independent and signature-based metagenomic
taxonomic profiling method with significantly smaller false discovery rates (FDR) that is 
laptop deployable. Our algorithm was tested and validated on twenty synthetic and mock 
datasets ranging in community composition and complexity, was applied successfully to data
generated from spiked environmental and clinical samples, and robustly demonstrates 
superior performance compared with other available tools.</p>

<hr>

<h3>
<a id="change-log" class="anchor" href="#change-log" aria-hidden="true"><span class="octicon octicon-link"></span></a>What's New
</h3>

<p>Some major changes has been made in the v1.0 release:
<li>The number of classified reads reports on the ouput in READ_COUNT column.</li>
<li>The number of split-trimmed reads that hit to plasmids reports in HIT_COUNT_PLASMID column.</li>
<li>GOTTCHA can now remove the plasmid hits using "--noPlasmidHit".</li>
<div style="padding:0.8em; background-color: #f9f8e6; border-color: rgba(231,228,157,.15); border-width: 5px; font-style: italic;">
Note: The plasmid relative results and the option need the new parsed database to work properly. For users who downloaded the databases before 30th March 2015, 
we encourage users to download the new parsed database (*.parsedGOTTCHA.dmp) to replace the old one. GOTTCHA v1.0 still supports old databases but 
plasmid relative results will show zero due to absence of the plasmid information.
</div>

</p>

<hr>

<h3>
<a id="system-requirement" class="anchor" href="#system-requirement" aria-hidden="true"><span class="octicon octicon-link"></span></a>Can my System Run GOTTCHA</h3>

<p>Linux (2.6 kernel or later) or Mac (OSX 10.6 Snow Leopard or later) operating system
with a minimum of 8 GB of RAM is recommended. Perl v5.8 or above is required. The C/C++
compiling enviroment might be required for installing dependencies. Systems may vary.
Please assure that your system has the essential software building packages (e.g. build-essential
for Ubuntu, XCODE for Mac...etc) installed properly before running the installing 
script.</p>

<p>GOTTCHA was tested successfully on our Linux servers (Ubuntu 12.10 w/ Perl v5.14.2; 
Ubuntu 10.04 w/ Perl v5.10.1) and Macbook Pro laptops (MAC OSX 10.8 w/ XCODE v5.1).</p>

<hr>

<h3>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span class="octicon octicon-link"></span></a>How to Install GOTTCHA</h3>

<p>The installation guide and the quick tutorial can be found in the the <a href="https://github.com/LANL-Bioinformatics/GOTTCHA">Github page</a>. More detail of each step is descripted in <a href="https://github.com/LANL-Bioinformatics/GOTTCHA/tree/master#detail-instructions">this section</a>.</p>

<hr>

<h3>
<a id="discussion" class="anchor" href="#discussion" aria-hidden="true"><span class="octicon octicon-link"></span></a>Discussion / Bug Report</h3>

<p>We have created a mailing list for GOTTCHA users. If you would like to recieve notifications about the updates and join the discussion, please join the mailing list by becoming the member of GOTTCHA-users groups.</p>
<p><a href="https://groups.google.com/d/forum/gottcha-users">GOTTCHA user's google group</a></p>
<p>Despite all these efforts, there are potential bugs and issues. Please help us to make it better by reporting them to GitHub issue tracker.</p>
<p><a href="https://github.com/LANL-Bioinformatics/GOTTCHA/issues">GOTTCHA issue tracker</a></p>
<p>Any other questions? You are welcome to contact Po-E (Paul) Li via po-e[at]lanl.gov.</p>

<hr>

<h3>
	<a id="quick-start" class="anchor" href="#quick-start" aria-hidden="true"><span class="octicon octicon-link"></span></a>I Can't Wait! How to Run GOTTCHA?</h3>

<p>This is an quick example of profiling a "test.fastq" file using GOTTCHA with a species-level
pre-computed bacterial database. The testing FASTQ file comes along with the GOTTCHA package
in the "test" directory. More details are stated in the INSTRUCTION section.</p>

<ol>
	<li>
	<p>Obtaining GOTTCHA package:</p>

	<pre><code>$ git clone https://github.com/LANL-Bioinformatics/GOTTCHA.git gottcha
	</code></pre>
	</li>
	<li>
	<p>Installing GOTTCHA:</p>

	<pre><code>$ cd gottcha
$ ./INSTALL.sh</code></pre>
	</li>
	<li>
	<p>Downloading lookup table and species-level database from our SFTP server.
	Use '9001gottcha' as the password when the request prompts:</p>

	<pre><code>$ sftp -o "Port 33001" gottcha@img-gp.lanl.gov:gottcha/GOTTCHA_lookup.tar.gz ./
$ sftp -o "Port 33001" gottcha@img-gp.lanl.gov:gottcha/GOTTCHA_BACTERIA_c3514_k24_u24.species.tar.gz ./</code></pre>

	<p>If you have any difficulty obtaining the databases, please contact us.</p>
	</li>
	<li>
	<p>Unpacking and decompressing archives:</p>

	<pre><code>$ tar -zxvf GOTTCHA_lookup.tar.gz
$ tar -zxvf GOTTCHA_BACTERIA_c3514_k24_u24.species.tar.gz</code></pre>
	</li>
	<li>
	<p>Running gottcha.pl: </p>

	<pre><code>$ bin/gottcha.pl              \
      --threads 8             \
      --outdir ./             \
      --input test/test.fastq \
      --database database/GOTTCHA_BACTERIA_c3514_k24_u24.species</code></pre>
	</li>
	<li><p>Enjoying the result at './test.gottcha.tsv'.</p></li>
</ol>
<h3>
	
<hr>

<a id="interpreting-results" class="anchor" href="#interpreting-results" aria-hidden="true"><span class="octicon octicon-link"></span></a>What's the Output</h3>

<p>GOTTCHA reports profiling results in a neat summary table (*.gottcha.tsv) by default.
The tsv file will list the organism(s) at all taxonomic levels from STRAIN to PHYLUM,
their linear length, total bases mapped, linear depth of coverage, and the normalized 
linear depth of coverage. The linear depth of coverage (LINEAR_DOC) is used to calculate 
relative abundance of each organism or taxonomic name in the sample.</p>

<p>Summary table:</p>

<table>
<thead>
<tr>
<th>Column</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>LEVEL</td>
<td>taxonomic rank</td>
</tr>
<tr>
<td>NAME</td>
<td>taxonomic name</td>
</tr>
<tr>
<td>REL_ABUNDANCE</td>
<td>relative abundance (equivalent to NORM_COV by default)</td>
</tr>
<tr>
<td>LINEAR_LENGTH</td>
<td>number of non-overlapping bases covering the signatures</td>
</tr>
<tr>
<td>TOTAL_BP_MAPPED</td>
<td>sum total of all hit lengths recruited to signatures</td>
</tr>
<tr>
<td>HIT_COUNT</td>
<td>number of hits recruited to signatures</td>
</tr>
<tr>
<td>HIT_COUNT_PLASMID</td>
<td>number of hits recruited to signatures</td>
</tr>
<tr>
<td>READ_COUNT</td>
<td>number of reads recruited to signatures</td>
</tr>
<tr>
<td>LINEAR_DOC</td>
<td>linear depth-of-coverage (TOTAL_BP_MAPPED / LINEAR_LENGTH)</td>
</tr>
<tr>
<td>NORM_COV</td>
<td>normalized linear depth-of-coverage (LINEAR_DOC / SUM(LINEAR_DOC in certain level))</td>
</tr>
</tbody>
</table>

<h3>
<a id="visulization" class="anchor" href="#visulization" aria-hidden="true"><span class="octicon octicon-link"></span></a>How to Visualize the Result</h3>

<p><a href="http://sourceforge.net/p/krona/home/krona/">Krona</a> is an interactive browser that allows 
the exploration of hierarchical data with pie charts. Assuming you have Krona installed properly,
we are going to create a Krona chart from a text file listing abundance and lineages. 
You will find .lineage.tsv file when you run gottcha.pl in "all" output mode.</p>

<p>Use 'ktImportText' and save the chart to "test.krona.html":</p>

<pre><code>    $ ktImportText test_temp/test.lineage.tsv -o test.krona.html
</code></pre>

<hr>

<h3>
<a id="citation" class="anchor" href="#citation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Citation</h3>

<p>Please cite GOTTCHA in your publications:</p>
<p>Tracey Allen K. Freitas, Po-E Li, Matthew B. Scholz and Patrick S. G. Chain (2015) <strong>Accurate read-based metagenome characterization using a hierarchical suite of unique signatures,</strong> Nucleic Acids Research (DOI: 10.1093/nar/gkv180)</p>

<p></p>

      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer" style="background: #212121 url('images/los-alamos-lab-logo-white.png') no-repeat fixed bottom; ">
      <footer class="inner">
	    <p class="copyright">GOTTCHA maintained by Bioscience Div, <a href="http://www.lanl.gov/">Los Alamos National Laboratory</a></p>
        <p>Published with <a href="https://pages.github.com">GitHub Pages</a></p>
		<p>This project is funded by U.S. Defense Threat Reduction Agency [R-00059-12-0].</p>
      </footer>
    </div>

    

  </body>
</html>

